src/shims/unix/fs.rs

```diff
@@ -1202,6 +1202,67 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         }
     }
 
+    /// NOTE: According to the man page of `possix_fallocate`, it returns the error code instead
+    /// of setting `errno`.
+    fn posix_fallocate(
+        &mut self,
+        fd_num: i32,
+        offset: i64,
+        len: i64,
+    ) -> InterpResult<'tcx, Scalar> {
+        let this = self.eval_context_mut();
+
+        // Reject if isolation is enabled.
+        if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
+            this.reject_in_isolation("`posix_fallocate`", reject_with)?;
+            // Return error code "EBADF" (bad fd).
+            return interp_ok(this.eval_libc("EBADF"));
+        }
+
+        let Some(fd) = this.machine.fds.get(fd_num) else {
+            return interp_ok(this.eval_libc("EBADF"));
+        };
+
+        let file = match fd.downcast::<FileHandle>() {
+            Some(file_handle) => file_handle,
+            // Man page specifies to return ENODEV if `fd` is not a regular file.
+            None => return interp_ok(this.eval_libc("ENODEV")),
+        };
+
+        // EINVAL is returned when: "offset was less than 0, or len was less than or equal to 0".
+        if offset < 0 || len <= 0 {
+            return interp_ok(this.eval_libc("EINVAL"));
+        }
+
+        if !file.writable {
+            // The file is not writable.
+            return interp_ok(this.eval_libc("EBADF"));
+        }
```

Comment:
The order in which you do these checks looks a bit random, let's clean that up
```suggestion
        // EINVAL is returned when: "offset was less than 0, or len was less than or equal to 0".
        if offset < 0 || len <= 0 {
            return interp_ok(this.eval_libc("EINVAL"));
        }

        // Get the file handle.
        let Some(fd) = this.machine.fds.get(fd_num) else {
            return interp_ok(this.eval_libc("EBADF"));
        };
        let file = match fd.downcast::<FileHandle>() {
            Some(file_handle) => file_handle,
            // Man page specifies to return ENODEV if `fd` is not a regular file.
            None => return interp_ok(this.eval_libc("ENODEV")),
        };
        if !file.writable {
            // The file is not writable.
            return interp_ok(this.eval_libc("EBADF"));
        }
```